packages/ra-core/src/auth/useLogout.ts

```diff
@@ -62,77 +62,81 @@ const useLogout = (): Logout => {
+                            pathname: redirectToParts[0],
+                        };
+                        let newLocationOptions = {};
 
-                if (
-                    redirectToCurrentLocationAfterLogin &&
-                    locationRef.current &&
-                    locationRef.current.pathname
-                ) {
-                    newLocationOptions = {
+                        if (
+                            redirectToCurrentLocationAfterLogin &&
+                            locationRef.current &&
+                            locationRef.current.pathname
+                        ) {
+                            newLocationOptions = {
+                                state: {
+                                    nextPathname: locationRef.current.pathname,
+                                    nextSearch: locationRef.current.search,
+                                },
+                            };
+                        }
+                        if (redirectToParts[1]) {
+                            newLocation.search = redirectToParts[1];
+                        }
+                        navigateRef.current(newLocation, newLocationOptions);
+                        resetStore();
+                        queryClient.clear();
+
+                        return redirectToFromProvider;
+                    });
+            } else {
+                navigate(
+                    {
+                        pathname: loginUrl,
+                    },
+                    {
                         state: {
-                            nextPathname: locationRef.current.pathname,
-                            nextSearch: locationRef.current.search,
+                            nextPathname: location && location.pathname,
                         },
-                    };
-                }
-                if (redirectToParts[1]) {
-                    newLocation.search = redirectToParts[1];
-                }
-                navigateRef.current(newLocation, newLocationOptions);
+                    }
+                );
                 resetStore();
                 queryClient.clear();
-
-                return redirectToFromProvider;
-            }),
-        [authProvider, resetStore, loginUrl, queryClient]
-    );
-
-    const logoutWithoutProvider = useCallback(
-        _ => {
-            navigate(
-                {
-                    pathname: loginUrl,
-                },
-                {
-                    state: {
-                        nextPathname: location && location.pathname,
-                    },
-                }
-            );
-            resetStore();
-            queryClient.clear();
-            return Promise.resolve();
+                return Promise.resolve();
+            }
         },
-        [resetStore, location, navigate, loginUrl, queryClient]
+        [authProvider, resetStore, loginUrl, queryClient, location, navigate]
```

Comment:
This is problematic: `location` changes on every navigation, so the callback memoization is weak. You're using a locationRef for the sarch and hash, why not for the pathName?

Reply:
I don't know why it was not done for `pathName` (I just moved the content of the second `useCallback`). But I assume we probably simply forgot to change this second callback to use refs while we changed the first callback, 2 years ago. Besides I can't think of any negative consequences this might have, so I used refs in this second part, too.
Nice catch!